### DISCUSSION TITLE
Upgrading BBN Versions

### DISCUSSION THREAD

**Discussion by aviljoenn** [bareboat-necessities/lysmarine_gen#314] [General]
What is the recommendation in terms of upgrading of BBN versions? E.g. I am currently running lysmarine-bbn-bullseye_2023-03-24-raspios-arm64.img.xz and I would like to go to lysmarine-bbn-bullseye_2023-06-15-raspios-arm64.img.xz
Can I upgrade my existing installation or must I re-install from scratch?
If there're docs or instructions about this, please feel free to point me in that direction?
Thank you so much.

**Comment by mgrouch** [comment]
If you have anything valuable you need to save it somewhere else than your SD card you going to be using. 

You load a new image from scratch. I do recommend 2023-06-15 as it fixes many things since your last one. OpenCPN performance with vector charts. 

I usually have 2 SD cards and one SSD. I rotate sd cards (one keeps previous backup). 
If you use rpi-clone to SSD, do not forget to delete all partitions from SSD before running rpi-clone. (Otherwise it copies over existing stuff without deleting)


There is something in documentation too. 
It’s in this section of the doc. I probably should move into upgrading notes. 

https://bareboat-necessities.github.io/my-bareboat/bareboat-os.html#_suggestions

The beauty of Linux is that you can customize it for your needs in infinite ways. While this distro aimed to strike common need, you will find that number of post-install customization steps would be required. The key is to script those steps, make them non-interactive, make the steps requite NO GUI. In that case your set-up becomes REPRODUCIBLE in case of new OS image releases. You can share your post-install scripts, so the system can be improved and even more fine-tuned.

**Reply by aviljoenn** [reply]
Thank you. Your last paragraph about my post-install customizations is indeed exactly why I asked the question. These include amongst others, my Python scripts, piped providers, drive mappings, custom ShipXplorer AIS stuff, openCPN maps, nmea0183 connections and more... You're right about the scripting. I think it's time that I put my head down and script all of this once and for all.
Thank you so much for your good work!

**Comment by kjsanders86** [comment]
Very helpful!  Has anyone added a post-install script that's been vetted?  One comprehensive enough to be added to the Commands menu?  Seems like we are always trying to keep our 'bits' updated to the latest version across all of our tech gear.  I do know there are many unique installs but standardization is very scalable!  

thx

**Reply by mgrouch** [reply]
Everyone has different set ups with different hardware. You can keep your own scripts somewhere in GitHub repository. 
Clone that repository post install and run scripts which will reduce manual repetitive tasks.

**Comment by atchik6056** [comment]
Hi Michael
just double check
I run BBN built from image 20230822.lysmarine.img.gz
which is bullseye 
```
user@lysmarine:~/add-ons $ cat /etc/*rel*
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
user@lysmarine:~/add-ons $ cat /etc/debian_version
11.8
```
I did a lot of customisation, including installation of SASplanet, qtvlm, modem manager for 4G connection for qualcomm device, few pi-apps, box64 and etc... turned bbn into router/firewall/access/ point and etc. It works great.
I can see now that the latest available bbn version is lysmarine-bbn-full-bookworm_2024-04-20-r1252-raspios-arm64 and I want to upgrade my install up to there.
if I understand recommendations I have to backup all my files, reinstall bbn from scratch using above image and reinstall all adds-ons and restore config files. This is bloody pain. This is why these days I prefer containers vs appliance like bbn, and ansible vs scripting, side note anyways...
Double checking debian standard upgrade procedure and raspberry specific flavour does not look promising either, although it is possible: https://www.debian.org/releases/bookworm/amd64/release-notes/ch-upgrading.en.html , https://raspberrytips.com/upgrade-raspberry-pi-os-bookworm/, and actually the last one directly warns about high probability of broken system as the result. However it is possible: https://gist.github.com/jauderho/6b7d42030e264a135450ecc0ba521bd8
also how about restoring ~/.config files from old image to save many settings for the apps? (perhaps except openbox config, although I did customisation here as well) 
Can you please comment on the above? Reinstall from scratch seems to be the only way, correct?
Thank you.

**Comment by atchik6056** [comment]
I am trying to upgrade  20230822.lysmarine to bookworm. I am following pretty mach this procedure https://raspberrytips.com/upgrade-raspberry-pi-os-bookworm/ and replaced /etc/apt/sources.list.d files with files listed in https://github.com/bareboat-necessities/lysmarine_gen/tree/bookworm/install-scripts/0-repositories/files. The upgrade went ok, except errors about incorrect repositories :
E: The repository 'http://ppa.launchpad.net/bbn-projects/bbn-kplex/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/bbn-projects/bbn-ppa/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/dansmith/chirp-snapshots/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Failed to fetch https://open-mind.space/repo/dists/bookworm-stable/InRelease  Clearsigned file isn't valid, got 'NOSPLIT' (does the network require authentication?)
E: The repository 'https://open-mind.space/repo bookworm-stable InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/lysmarine/upstream-projects/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/lysmarine/rtl-ais/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/lysmarine/pypilot/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ppa.launchpad.net/openplotter/openplotter/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.

are they incorrect files/repo config  for bookworm? there is no jammy repo in repositories above.
missing something?

**Reply by mgrouch** [reply]
You won’t be able to upgrade it this way

**Reply by atchik6056** [reply]
fine, but this is not the answer. so I am just playing using available info from your github 
what is the right way to upgrade then? There must be a way to upgrade with preserving your data, settings and etc relevant to your boat as much as you can. Otherwise it wouldn't make sense to use for real life
what about seatalk settings which is a core communication hub for BBN?
If the reinstall is the answer, and then a pain reinstalling all additional apps, then it should be a way  to preserve settings (user and system) and data (perhaps only user) . Apparently not everything in /home/user can be moved over. What about system settings in /etc? for example network, vpn, firewall...?
Is there any recommendations or documented process? Might be not ideal but at least something?

**Reply by mgrouch** [reply]
The upgrade between major releases is not really supported by Debian or Raspberry. 
Fir good reasons. It’s untestable. 
You should have scripted your configuration changes (and kept those scripts somewhere on github or else) so you could apply them to a newly loaded image. Or at least kept your notes so you could easily repeat your customizations.

**Reply by atchik6056** [reply]
understood
what about user data, maps, settings in /home/user/.config for the apps?  
Network settings, firewall rules ?
PS: for now I ended up with working system but no desktop, although all other apps seem to be working ok, all data is preserved
![Screenshot_2024-06-11_13-19-23](https://github.com/bareboat-necessities/lysmarine_gen/assets/34929784/dc0550c0-1a74-4622-a458-48e710c72abd)

**Reply by mgrouch** [reply]
Back up them on external drive or keep whole SD card and use new SD card for new image. 
When settings up new system you can look and copy things from old one.

**Reply by atchik6056** [reply]
just my two cents
this is kinda trivial for linuxoids but for non linux people , even for windows guys, it is not so easy
would be nice to have some guidance or overview  what is what and what to copy, links, permissions, basic settings like network to move over
it would help to non experienced people
PS: I gave up with direct upgrade, as apparently there are too many inconsistencies between bullseye and bookworm desktops
I do not know why you did not mentioned earlier that the repositories I mentioned above with errors are not even included into /etc/apt/sources.list.d in your image. so my attempt of direct upgrade  was actually successful except the desktop itself - which is a big issue though. I tried to fix it with raspi-config or reinstalling onbox LXD - no luck. Wish to have more docs on bookworm vs bullseye... internet info sucks these days
anyways I am not particularly happy with bbn upgrade status quo

**Reply by mgrouch** [reply]
It is in the documentation under section “Suggestions”. 

Quote:

````
The beauty of Linux is that you can 
customize it for your needs in infinite 
ways. While this distro aimed to strike 
common need, you will find that number 
of post-install customization steps would 
be required. The key is to script those 
steps, make them non-interactive, make 
the steps require NO GUI. In that case 
your set-up becomes REPRODUCIBLE in 
case of new OS image releases. You can 
share your post-install scripts, so the 
system can be improved and even more 
fine-tuned.
````

BTW your desktop is not LXD. It’s Budgie Desktop. Also in the documentation.

**Reply by atchik6056** [reply]
the above is just generic statement
i think you can see what i meant if you read my comments above 
upgrade process and documentation is developer responsibility, if it does not work or cumbersome - people just go away,
yes this is budgie, my typo

**Reply by mgrouch** [reply]
> the above is just generic statement i think you can see what i meant if you read my comments above upgrade process and documentation is developer responsibility, if it does not work or cumbersome - people just go away, yes this is budgie, my typo

There is no point in documentation if it’s not read by users. It seem you are included. Upgrade process the way you expect it requires all components to support it smoothly and that doesn’t exist. If you are trying to say that it’s my responsibility you are simply wrong.

**Reply by atchik6056** [reply]
it is too bad you are attacking your customer. You missed my point. You documentation is actually very good for open source project. I am just saying that it is missing upgrade procedure, except very generic statement about customising new linux install **from scratch**. Perhaps the reason is that It is rather an appliance which is typically not up-gradable between major releases. If so it would nice to have a preserved user space, environment and data, customise apps to keep data in there or document it at least. and of course this is not your fault or problem that debian made unbelievable mess going from bullseye to bookworm desktop.
